MoltenVK/MoltenVK/Commands/MVKCmdQueries.mm

```diff
@@ -117,12 +117,13 @@
 }
 
 void MVKCmdCopyQueryPoolResults::encode(MVKCommandEncoder* cmdEncoder) {
-    [cmdEncoder->_mtlCmdBuffer addCompletedHandler: ^(id<MTLCommandBuffer> mtlCmdBuff) {
-        // This can block, so it must not run on the Metal completion queue.
-        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
-            _queryPool->copyQueryPoolResults(_query, _queryCount, _destBuffer, _destOffset, _destStride, _flags);
-        });
-    }];
+    // What happens now depends on whether or not I was added before or after the query ended.
+    if (!_queryPool->areQueriesDeviceAvailable(_query, _queryCount) && mvkIsAnyFlagEnabled(_flags, VK_QUERY_RESULT_WAIT_BIT)) {
+        // Defer this until the queries will be done.
+        _queryPool->deferCopyResults(_query, _queryCount, _destBuffer, _destOffset, _destStride, _flags);
+    } else {
+        _queryPool->encodeCopyResults(cmdEncoder, _query, _queryCount, _destBuffer, _destOffset, _destStride, _flags);
+    }
 }
```

Comment:
I'm afraid I've lost track of the activity order.

What is it we are now protecting when this deferral is triggered? 

Since the copy now happens on the GPU, can't we assume the results are available (at least on the GPU), and just let the buffer copy happen, without waiting for the command buffer execution to be completed to provide the availability feedback to the CPU?

Reply:
> What is it we are now protecting when this deferral is triggered?
> 
> Since the copy now happens on the GPU, can't we assume the results are available (at least on the GPU), and just let the buffer copy happen, without waiting for the command buffer execution to be completed to provide the availability feedback to the CPU?

But we're not waiting for command buffer to be *executed*, merely for the commands to finish all the queries to be *encoded*. *That's* what this is waiting for. When the last of the `vkCmdEndQuery()` commands is encoded, then the query object will encode the command to copy the results out of the buffer.

Reply:
Ah. Shoot...I was forgetting about `VK_QUERY_RESULT_WAIT_BIT`.

Okay. Thanks for clarifying. Makes supreme sense. (Although I must say the Vulkan spec could have saved implementations a bunch of trouble by requiring clients to end queries before requesting their results!)

Very elegant implementation!